plugins/backend-ekg/src/Cardano/BM/Backend/Prometheus.lhs

```diff
@@ -7,39 +7,66 @@
 
 module Cardano.BM.Backend.Prometheus
     ( spawnPrometheus
-    , passToPrometheus
     ) where
 
 import qualified Control.Concurrent.Async as Async
 import           Control.Monad.IO.Class (MonadIO (..))
-import           Network.Wai.Handler.Warp (Port, HostPreference, defaultSettings, runSettings,
-                   setHost, setPort)
-import           System.Metrics.Prometheus.Http.Scrape (prometheusApp)
-import           System.Metrics.Prometheus.Registry (RegistrySample, sample)
-import           System.Metrics.Prometheus.RegistryT (execRegistryT)
+import qualified Data.HashMap.Strict as HM
+import           Data.ByteString.Builder
+import           Data.ByteString.Char8 (ByteString)
+import           Data.Int (Int64)
+import           Data.Text (Text, replace)
+import           Data.Text.Encoding (encodeUtf8)
+import           Snap.Core (Snap, route, writeLBS)
+import           Snap.Http.Server (Config, ConfigLog (..), defaultConfig, setAccessLog,
+                     setErrorLog, setHostname, setPort, simpleHttpServe)
+import           System.Metrics (Value (..), sampleAll)
 import qualified System.Remote.Monitoring as EKG
-import           System.Remote.Monitoring.Prometheus (registerEKGStore,
-                   AdapterOptions (..))
 
 \end{code}
 %endif
 
 \subsubsection{Spawn Prometheus client from existing EKG server}
 \label{code:spawnPrometheus}\index{spawnPrometheus}
-\label{code:passToPrometheus}\index{passToPrometheus}
 \begin{code}
 
-spawnPrometheus :: EKG.Server -> HostPreference -> Port -> IO (Async.Async ())
-spawnPrometheus s h p = Async.async $ passToPrometheus s h p
-
-passToPrometheus :: EKG.Server -> HostPreference -> Port -> IO ()
-passToPrometheus server host port =
-    let store = EKG.serverMetricStore server
-        reg = execRegistryT $ registerEKGStore store $ AdapterOptions mempty Nothing 1
-    in serveMetrics (reg >>= sample)
+spawnPrometheus :: EKG.Server -> ByteString -> Int -> IO (Async.Async ())
+spawnPrometheus ekg host port = Async.async $
+    simpleHttpServe config site
   where
-    serveMetrics :: MonadIO m => IO RegistrySample -> m ()
-    serveMetrics = liftIO . runSettings settings . prometheusApp ["metrics"]
-    settings = setPort port . setHost host $ defaultSettings
+    config :: Config Snap a
+    config = setPort port . setHostname host . setAccessLog ConfigNoLog . setErrorLog ConfigNoLog $ defaultConfig
+    site :: Snap ()
+    site = route [
+             ("/metrics/", webhandler ekg)
+           ]
+
+webhandler :: EKG.Server -> Snap ()
+webhandler ekg = do
+    samples <- liftIO $ sampleAll $ EKG.serverMetricStore ekg
+    writeLBS . toLazyByteString . renderSamples $ HM.toList samples
+    pure ()
 
+renderSamples :: [(Text, Value)] -> Builder
+renderSamples [] = mempty
```

Comment:
This is kind of manual, but I guess for this level of output, it's ok?
It won't be if we start adding more output.

Reply:
thanks for your inputs. I'll make it nicer..